### PR TITLE
Implement algorithm to better classify ship pings

### PR DESCRIPTION
**Pull request by andyphelps**
* Will interpolate pings appropriately and normalise data so that pings are at standard rate
* Add criteria where no interpolation is required (> 6 hour gap or 30km gap)
* Add type for ShipPings to describe filtered pings

### CODE REVIEW

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package uk.gov.ukho.ais.rasters
+
+import java.sql.Timestamp
+import java.time.Instant
+
+case class ShipPing(
+    final val mmsi: String,
```
Comment (JoelClemence) [inline spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala]: Can this be removed now we have the algorithm down?

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    final val acquisitionTime: Long,
+    final val latitude: Double,
+    final val longitude: Double
+) {
+  override def toString: String =
+    s"ShipPing[$mmsi,${Timestamp.from(Instant.ofEpochMilli(acquisitionTime))}, $longitude, $latitude]"
```
Comment (JoelClemence) [inline spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala]: Is this needed now we have the algorithm in place?

Comment (JoelClemence) [reply spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala]: Although, would help with test failures

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/Interpolator.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        prevPing = currPing
+      })
+
+    outputPings
+  }
+
+  private def interpolateBetweenPingsFromTime(prevPing: ShipPing,
+                                              currPing: ShipPing,
+                                              fromTime: Long): List[ShipPing] =
+    if (shouldInterpolateBetweenPings(prevPing, currPing)) {
```
Comment (JoelClemence) [inline spark-job/src/main/scala/uk/gov/ukho/ais/rasters/Interpolator.scala]: Not sure method name is correct - ```shouldNotInterpolateBetweenPings``` or ```shouldSkipInterpolationBetweenPings``` or the like would be better in this case or invert the cases and correct the logic in the method

File: spark-job/README.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (JoelClemence) [inline spark-job/README.md]: Do we want to expand this to describe the alogorithm?

Comment (ciaranevans) [reply spark-job/README.md]: Can we just screenshot the page from the document?

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/AisToRaster.scala [hunk 5]
```diff
@@ -47,17 +48,30 @@ object AisToRaster {
...
+    val shipPings = spark.read
       .schema(Schema.AIS_SCHEMA)
       .option("sep", "\t")
       .csv(config.inputPath)
 
-    val geoPoints = filterShipPings(shipPoints)
+    val filteredShipPings = filterShipPingsByMessageType(shipPings)
 
-    geoPoints
+    // Normalise ping frequencies for each MMSI
```
Comment (ciaranevans) [inline spark-job/src/main/scala/uk/gov/ukho/ais/rasters/AisToRaster.scala]: Commented lines can be removed

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/AisToRaster.scala [hunk 6]
```diff
@@ -96,13 +110,19 @@ object AisToRaster {
     s"$prefix-raster-$timestamp"
```
Comment (ciaranevans) [inline spark-job/src/main/scala/uk/gov/ukho/ais/rasters/AisToRaster.scala]: Could we mention this is the full contents of the s3 bucket for now in the filename?

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/Interpolator.scala (hunk 3 above)
Comment (ciaranevans) [reply spark-job/src/main/scala/uk/gov/ukho/ais/rasters/Interpolator.scala]: Agree ^ Naming seems to contradict the behaviour

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala (hunk 1 above)
Comment (ciaranevans) [reply spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala]: I'd prefer keeping it to be honest, easier to read IMO.

File: spark-job/src/test/scala/uk/gov/ukho/ais/rasters/InterpolatorTest.scala [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  @Test
+  def whenInterpolationOfZeroLengthSequenceThenZeroLengthSequenceReturned()
+    : Unit = {
+    val outPings: Seq[ShipPing] = Interpolator.interpolatePings(Seq[ShipPing]())
+
+    assertThat(outPings.size).isEqualTo(0)
+  }
+
+  @Test
+  def whenInterpolationOfLengthOf1SequenceThenZeroLengthSequenceReturned()
```
Comment (ciaranevans) [inline spark-job/src/test/scala/uk/gov/ukho/ais/rasters/InterpolatorTest.scala]: Incorrect name - whenInterpolationOfLengthOf1SequenceThen1LengthSequenceReturned()

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala (hunk 2 above)
Comment (oliverCausley) [reply spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala]: We might as well leave it in  - could be handy in the future

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala (hunk 1 above)
Comment (andyphelps) [reply spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala]: Agreed to keep

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala (hunk 2 above)
Comment (andyphelps) [reply spark-job/src/main/scala/uk/gov/ukho/ais/rasters/ShipPing.scala]: I would prefer to keep it for that very reason

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/Interpolator.scala (hunk 3 above)
Comment (andyphelps) [reply spark-job/src/main/scala/uk/gov/ukho/ais/rasters/Interpolator.scala]: Will fix.

File: spark-job/src/test/scala/uk/gov/ukho/ais/rasters/InterpolatorTest.scala (hunk 7 above)
Comment (andyphelps) [reply spark-job/src/test/scala/uk/gov/ukho/ais/rasters/InterpolatorTest.scala]: Done

File: spark-job/src/main/scala/uk/gov/ukho/ais/rasters/AisToRaster.scala (hunk 6 above)
Comment (andyphelps) [reply spark-job/src/main/scala/uk/gov/ukho/ais/rasters/AisToRaster.scala]: Needs to be done in the lambda.  Will fix it in later commit

File: spark-job/README.md (hunk 4 above)
Comment (andyphelps) [reply spark-job/README.md]: In another pbi.

**Pull request merged**